src/base/bitarray.h

```diff
@@ -0,0 +1,62 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2021  Vladimir Golovnev <glassez@yandex.ru>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#pragma once
+
+#include <libtorrent/bitfield.hpp>
+
+class BitArray
```

Comment:
I think this class should live under `bittorrent` namespace/folder. For other non-related cases, I would prefer `QBitArray`.

Reply:
It is not directly related to BitTorrent, so I would prefer to have it outside of this namespace. This doesn't stop you from using QBitArray wherever you want.

Reply:
> For other non-related cases, I would prefer `QBitArray`.

I don't see any objective meaning in this. "bit array" is trivial enough to prefer one of the implementations. The only reason I (long ago) decided to use libtorrent implementation is to avoid meaningless conversions from one array to another. In addition, using it through our own wrapper, we are free from real implementation. We can change it if necessary, without affecting the callers any more.

Reply:
>I don't see any objective meaning in this.

Qt is a platform toolkit while libtorrent is a library for a specific protocol. In general, I'm inclined to think that a library is free to only provide the bits it needs and not aiming to be as generic as possible. I can't tell whether libtorrent will move bitfield.hpp into "_aux" namespace in the future.

I'm a bit concerned that we have too many implementations at hand: libtorrent, Qt , the new wrapper here. It is not easy for a newcomer (or someone doesn't follow latest development) to see which one is preferred (or why preferring it). Therefore my suggestion to move it into `bittorrent` namespace to make it specific/obvious. However if you don't think it is a real problem, then so be it.

Reply:
>I can't tell whether libtorrent will move bitfield.hpp into "_aux" namespace in the future.

How can it hide what is part of its interface? This is not some class that is used purely internally.
>I'm a bit concerned that we have too many implementations at hand: libtorrent, Qt , the new wrapper here.

There's only two implementations. My BitArray implements nothing. It is used to provide implementation independent interface for other classes.
In addition, the presence of two available implementations is not our whim. This is just a consequence of the fact that we use several different libraries/frameworks that provide it.
>It is not easy for a newcomer (or someone doesn't follow latest development) to see which one is preferred (or why preferring it).

In my opinion, it is obvious that you should prefer project-specific things, if they exist. Otherwise, why do they exist?

In fact, I wouldn't bother with this if there was a more optimal way of converting from libtorrent bitfield to QBitArray than bitwise copying.

Reply:
>if there was a more optimal way of converting from libtorrent bitfield to QBitArray than bitwise copying.

Well, time doesn't stand still... It looks like we can just use `QBitArray::fromBits()` starting with the v4.4. So I'll try it out first.

Reply:
It looks like they have a different layout after all, so this won't work.

Reply:
>Therefore my suggestion to move it into `bittorrent` namespace to make it specific.

But it has nothing to do with bittorrent at all. It's just an array of bits.

Reply:
>But it has nothing to do with bittorrent at all. It's just an array of bits.

It is a bitfield implementation used in a bittorrent library...

Reply:
>>But it has nothing to do with bittorrent at all. It's just an array of bits.
>
>It is a bitfield implementation used in a bittorrent library...

No, it isn't. We're not talking about `lt::bitfield`. We're talking about qBittorrent native class that provides its users with a "bit array" interface that has nothing to do with BitTorrent (it's just array of booleans). What it uses `lt::bitfield` internally is the implementation details. Today they are the same, tomorrow they may change, but this will not affect the interface.

Reply:
>No, it isn't. We're not talking about lt::bitfield. 

Yes we are.

>What it uses lt::bitfield internally is the implementation details. 

That contradicts with your goal of this very PR, the implementation details is tied with the performance gains so it is relevant with our discussion here.

>>Therefore my suggestion to move it into bittorrent namespace to make it specific.
>
>But it has nothing to do with bittorrent at all. It's just an array of bits.

Anyway I don't see discussion is going anywhere, I already give the approve, so whatever.